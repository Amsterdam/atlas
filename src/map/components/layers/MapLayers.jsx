import React from 'react';
import PropTypes from 'prop-types';

import MapLegend from '../legend/MapLegend';
import Notification from '../../../shared/components/notification/Notification';
import AddIcon from '../../../../public/images/icon-plus.svg';
import RemoveIcon from '../../../../public/images/icon-cross.svg';

import './_map-layers.scss';

const showLayer = (layer, user) => (
  !layer.authScope || (user.authenticated && user.scopes.includes(layer.authScope))
);

const showCategory = (layers, user) => (
  layers.filter((layer) => layer.category && showLayer(layer, user))
);

const MapLayers = (props) => (
  <div className="map-layers">
    <h3 className="u-sr-only">Beschikbare kaartlagen</h3>
    <ul>
      {[...new Set(showCategory(props.layers, props.user).map((layer) => layer.category))]
        .map((category) => (
          <li className="map-layers__category" key={category}>
            <h4 className="map-layers__category-text">{category}</h4>
            <ul>
              {props.layers
                .filter((layer) => layer.category === category && showLayer(layer, props.user))
                .map((layer) => (
                  <li
                    className={`
                      map-layers__title
                      map-layers__title--${props.activeMapLayers.some((mapLayer) => layer.title === mapLayer.title) ? 'active' : 'inactive'}
                    `}
                    key={layer.title}
                  >
                    <button onClick={() => {
                      MapLegend.mapLayersLegendItemsToIds(layer).forEach((mapLayerId) =>
                        props.onLayerToggle(mapLayerId));
                    }}
                    >
                      <span className="map-layers__toggle-title">
                        {layer.title}
                      </span>
                      <span className="map-layers__toggle map-layers__toggle--remove">
                        <RemoveIcon />
                      </span>
                      <span className="map-layers__toggle map-layers__toggle--add">
                        <AddIcon />
                      </span>
                    </button>
                  </li>
              ))}
            </ul>
          </li>
      ))}
    </ul>
    {!props.user.authenticated && (
      <Notification>
        <a
          href="#?mpb=topografie&pgn=content-overzicht&pgi=item1&pgt=inloggen"
        >Meer kaartlagen</a> na inloggen.
      </Notification>
    )}
  </div>
);

MapLayers.propTypes = {
  activeMapLayers: PropTypes.array, // eslint-disable-line
  layers: PropTypes.array, // eslint-disable-line
  onLayerToggle: PropTypes.func.isRequired, // eslint-disable-line react/no-unused-prop-types
  user: PropTypes.object // eslint-disable-line
};

export default MapLayers;
