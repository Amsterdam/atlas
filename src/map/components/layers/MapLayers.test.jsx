import React from 'react'
import { shallow } from 'enzyme'
import { useMatomo } from '@datapunt/matomo-tracker-react'
import MapLayers from './MapLayers'

jest.mock('@datapunt/matomo-tracker-react')

const panelLayers = [
  {
    id: 'id1',
    title: 'foo',
    mapLayers: [
      {
        id: 'id3',
        layers: ['layer'],
        legendItems: [
          {
            title: 'Legend 1',
          },
          {
            title: 'Legend 2',
          },
        ],
        title: 'Foo Layer',
        url: '/maps/layer',
        detailUrl: 'geosearch/search/',
        detailItem: 'foo',
      },
      {
        id: 'id2',
        layers: ['layer'],
        legendItems: [
          {
            title: 'Legend 1',
          },
          {
            title: 'Legend 2',
          },
        ],
        maxZoom: 16,
        minZoom: 8,
        title: 'Foo Layer',
        url: '/maps/layer',
        noDetail: true,
      },
    ],
  },
]

const activeLayers = [panelLayers[0]]

describe('MapLayer', () => {
  const mockTrackEvent = jest.fn()

  beforeEach(() => {
    useMatomo.mockImplementationOnce(() => ({ trackEvent: mockTrackEvent }))
  })

  it('should render', () => {
    const component = shallow(
      <MapLayers
        activeMapLayers={activeLayers}
        panelLayers={panelLayers}
        onLayerToggle={() => {}}
      />,
    )

    expect(component).toMatchSnapshot()
  })

  it('should render without active layers', () => {
    const component = shallow(
      <MapLayers activeMapLayers={[]} panelLayers={panelLayers} onLayerToggle={() => {}} />,
    )

    expect(component).toMatchSnapshot()
  })

  it('should call onLayerToggle when user clicks on button', () => {
    const onLayerToggleMock = jest.fn()

    const component = shallow(
      <MapLayers
        activeMapLayers={activeLayers}
        panelLayers={panelLayers}
        onLayerToggle={onLayerToggleMock}
      />,
    )

    const button = component.find('button').at(0)
    button.simulate('click')

    const firstCollection = panelLayers[0]
    const firstLayer = firstCollection.mapLayers[0]

    expect(onLayerToggleMock).toHaveBeenCalledWith(firstLayer)

    expect(mockTrackEvent).toHaveBeenCalledWith({
      category: 'kaartlaag',
      action: firstCollection.title,
      name: firstLayer.title,
    })
  })
})
