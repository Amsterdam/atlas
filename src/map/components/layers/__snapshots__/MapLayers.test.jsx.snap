// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapLayer should render 1`] = `
<div
  className="map-layers"
>
  <h3
    className="u-sr-only"
  >
    Beschikbare kaartlagen
  </h3>
  <ul>
    <li
      className="map-layers__category"
      key="id1"
    >
      <h4
        className="map-layers__category-text"
      >
        foo
      </h4>
      <ul>
        <li
          className="
                      map-layers__title
                      map-layers__title--inactive
                    "
          key="Foo Layer"
        >
          <button
            onClick={[Function]}
            type="button"
          >
            <span
              className="map-layers__toggle-title"
            >
              Foo Layer
            </span>
            <span
              className="map-layers__toggle map-layers__toggle--remove"
            />
            <span
              className="map-layers__toggle map-layers__toggle--add"
            />
          </button>
        </li>
        <li
          className="
                      map-layers__title
                      map-layers__title--inactive
                    "
          key="Foo Layer"
        >
          <button
            onClick={[Function]}
            type="button"
          >
            <span
              className="map-layers__toggle-title"
            >
              Foo Layer
            </span>
            <span
              className="map-layers__toggle map-layers__toggle--remove"
            />
            <span
              className="map-layers__toggle map-layers__toggle--add"
            />
          </button>
        </li>
      </ul>
    </li>
  </ul>
</div>
`;

exports[`MapLayer should render without active layers 1`] = `
<div
  className="map-layers"
>
  <h3
    className="u-sr-only"
  >
    Beschikbare kaartlagen
  </h3>
  <ul>
    <li
      className="map-layers__category"
      key="id1"
    >
      <h4
        className="map-layers__category-text"
      >
        foo
      </h4>
      <ul>
        <li
          className="
                      map-layers__title
                      map-layers__title--inactive
                    "
          key="Foo Layer"
        >
          <button
            onClick={[Function]}
            type="button"
          >
            <span
              className="map-layers__toggle-title"
            >
              Foo Layer
            </span>
            <span
              className="map-layers__toggle map-layers__toggle--remove"
            />
            <span
              className="map-layers__toggle map-layers__toggle--add"
            />
          </button>
        </li>
        <li
          className="
                      map-layers__title
                      map-layers__title--inactive
                    "
          key="Foo Layer"
        >
          <button
            onClick={[Function]}
            type="button"
          >
            <span
              className="map-layers__toggle-title"
            >
              Foo Layer
            </span>
            <span
              className="map-layers__toggle map-layers__toggle--remove"
            />
            <span
              className="map-layers__toggle map-layers__toggle--add"
            />
          </button>
        </li>
      </ul>
    </li>
  </ul>
</div>
`;
